xchange-stream-bitmex/src/main/java/info/bitrich/xchangestream/bitmex/BitmexStreamingMarketDataService.java

```
@@ -48,33 +53,72 @@ private String getBitmexSymbol(CurrencyPair currencyPair) {
+          if (orderBook != null) {
+            orderBook.getAsks().forEach(r -> {
+              LimitOrder order = new LimitOrder.Builder(ASK, currencyPair)
+                .originalAmount(r.get(1))
+                .limitPrice(r.get(0))
+                .build();
+              asks.add(order);
+            });
+            orderBook.getBids().forEach(r -> {
+              LimitOrder order = new LimitOrder.Builder(BID, currencyPair)
+                .originalAmount(r.get(1))
+                .limitPrice(r.get(0))
+                .build();
+              bids.add(order);
+            });
+          }
+          return new OrderBook(new Date(), asks, bids);
+        });
+    } else {
+      return streamingService
         .subscribeBitmexChannel(channelName)
         .map(
-            s -> {
-              BitmexOrderbook orderbook;
-              String action = s.getAction();
-              if (action.equals("partial")) {
-                orderbook = s.toBitmexOrderbook();
-                orderbooks.put(instrument, orderbook);
-              } else {
-                orderbook = orderbooks.get(instrument);
-                // ignore updates until first "partial"
-                if (orderbook == null) {
-                  return new OrderBook(
-                      new Date(), Collections.emptyList(), Collections.emptyList());
-                }
-                BitmexLimitOrder[] levels = s.toBitmexOrderbookLevels();
-                orderbook.updateLevels(levels, action);
+          s -> {
+            BitmexOrderbook orderbook;
+            String action = s.getAction();
+            if (action.equals("partial")) {
+              orderbook = s.toBitmexOrderbook();
+              orderbooks.put(instrument, orderbook);
+            } else {
+              orderbook = orderbooks.get(instrument);
+              // ignore updates until first "partial"
+              if (orderbook == null) {
+                return new OrderBook(
```

will we throw is we just return null instead of allocating empty orderbook structure? if not then just return null

---

I think it may affects external systems logic (actually it is not my code, may touch it with formatting), so i think this should be not changed